### DISCUSSION TITLE
`UI.Handle` vs `Hierarchy`

### DISCUSSION THREAD

**Discussion by laultman** [StereoKit/StereoKit#452] [General]
I'm creating a model with lots of individual objects that I want to manipulate as a single model, a building for example with all its individual object components.
Should I use a UI.Handle or a Hierarchy? If I wanted to have several such models (say floors of a building), what are the best options and what are the trade-offs.
They both seem to group a collection of things together.

**Comment by maluoi** [comment]
`UI.Handle` and `Hierarchy` do share some similarities, but let's refine this to a more direct equivalent! `UI.Handle` calls `UI.PushSurface`, which is UI hierarchy functionality without the grabbing/moving. This is the UI's direct equivalent of `Hierarchy.Push`!

In actuality, `UI.PushSurface` _calls_ `Hierarchy.Push`! It also makes some internal UI information available for that `Hierarchy` level, and kicks off some layout data, so it's a Hierarchy **_and_** UI.

Rule of thumb is, if you're making calls to `UI` functions, it should be inside of a `UI.PushSurface/UI.Handle/UI.Window`. If all you've got inside is draw calls and non-UI code, you can use a `Hierarchy.Push` instead. `UI.Handle` and such are also fine to have inside of a `Hierarchy.Push/Pop`!

Hope that's insightful :)

**Comment by laultman** [comment]
That explanation was very insightful indeed. My building analogy interpretation would seem to benefit from a hierarchy for all the building structural components (columns, walls, door openings, etc.) that are static immovable in the building space. The furniture in a room could each have its UI.Handle for manipulation. Finally, I could wrap the whole thing in a UI bounding box when I want a mode that contains the whole thing. Naturally, I could enter a mode where the bounding box is not there for interactions with the furniture elements.
